src/TestResults/match-by-context.ts

```diff
@@ -0,0 +1,378 @@
+      list.push(n.data);
+    } else {
+      list.push(...n.nodes.map((nn) => nn.data));
+    }
+    return list;
+  }, []);
+
+type MessageType = 'context-mismatch' | 'match-failed' | 'unusual-match';
+const makeWarning = (fileName: string) => (
+  messageType: MessageType,
+  contextType: ContextType,
+  source: ContainerNode<ItBlock> | DataNode<ItBlock>,
+  assertion?: ContainerNode<TestAssertionStatus> | DataNodeType<TestAssertionStatus>,
+  extraReason?: string
+): void => {
+  const output = (message: string): void =>
+    console.warn(`[${fileName}] ${message} \n source=`, source, `\n assertion=`, assertion);
+
+  const blockType = contextType === 'container' ? 'describe' : 'test';
+  switch (messageType) {
+    case 'context-mismatch':
+      output(
+        `!! context mismatched !! ${contextType} nodes are different under "${source.name}": `
+      );
+      break;
+    case 'match-failed': {
+      output(`!! match failed !! ${blockType}: "${source.name}": ${extraReason} `);
+      break;
+    }
+    case 'unusual-match': {
+      output(`unusual match: ${extraReason} : ${blockType} ${source.name}: `);
+      break;
+    }
+  }
+};
+
+/**
+ * match tests container with assertion container by their context structure.
+ *
+ * @param tContainer
+ * @param aContainer
+ * @param isGreedy optional, default to false. If the context did not match, we can still
+ * try to match by name. However it is not clear what could cause context misalignment in real world,
+ * therefore, hard to decide if we even need this?
+ */
+
+export const matchByContext = (
+  fileName: string,
+  tContainer: ContainerNode<ItBlock>,
+  aContainer: ContainerNode<TestAssertionStatus>
+): TestResult[] => {
+  const warning = makeWarning(fileName);
+
+  const _matchByContext = (
+    _tContainer: ContainerNode<ItBlock>,
+    _aContainer: ContainerNode<TestAssertionStatus>
+  ): TestResult[] => {
+    /* simple name-based look up as a fallback, best effort attempt for mismatched data nodes */
+    const matchDataByName = (): TestResult[] => {
+      let assertions = extractContainerData(_aContainer);
+      const results = _tContainer.childData.map((node) => {
+        const n = assertSourceDataNode(node);
+        const testName = n.data.name;
+        const found = assertions.filter((a) => a.title === testName);
+        switch (found.length) {
+          case 0: {
+            const reason = `missing assertion`;
+            warning('match-failed', 'data', n, undefined, reason);
+            return toMatchResult(n.data, reason);
+          }
+          case 1:
+            assertions = assertions.filter((a) => a != found[0]);
+            return toMatchResult(n.data, found[0]);
+          default: {
+            const reason = `found too many (${found.length}) assertions`;
+            warning('match-failed', 'data', n, _aContainer, reason);
+            return toMatchResult(n.data, reason);
+          }
+        }
+      });
+      return results;
```

Comment:
Is it possible that at this point there are still some unmatched `assertions` left? Should we issue a warning in this case?

Reply:
no need, see earlier comment...